lib/bindings/bmap.c

```diff
@@ -0,0 +1,251 @@
+            /* (at) is not on a word boundary; shift so we can use ffsl */
+            bit = __builtin_ffsl(~bmap[i] >> shift);
+        else
+            bit = __builtin_ffsl(~bmap[i]);
+        if (bit)
+            break;
+    }
+
+    if (bit) {
+        if (i == word)
+            /* Restore previous shift if any */
+            bit += shift;
+        return (i * BPW) + (bit - 1);
+    }
+    else
+        return -1;
+}
+
+/*
+ * Set (n) bits in (bmap[]) at 0-based bit index (at).
+ */
+static void setn_at(long *bmap, size_t bmap_size, int at, int n)
+{
+    assert((at + n - 1) < (bmap_size * BPW));
+    while (n > 0) {
+        n -= 1;
+        bmap[((at + n) / BPW)] |= (1UL << ((at + n) % BPW));
+    }
+}
+
+/*
+ * Clear (n) bits in (bmap[]) at 0-based bit index (at).
+ */
+static void clearn_at(long *bmap, size_t bmap_size, int at, int n)
+{
+    assert((at + n - 1) < (bmap_size * BPW));
+    while (n > 0) {
+        n -= 1;
+        bmap[((at + n) / BPW)] &= ~(1UL << ((at + n) % BPW));
+    }
+}
+
+/*
+ * Allocate (n) pages from (alloc), returns a memory address or NULL if no
+ * space found.
+ */
+void *bmap_alloc(bmap_allocator_t *alloc, size_t n)
+{
+    int a = 0, b = 0;
+    size_t bmap_bits = alloc->bmap_size * BPW;
+
+    /*
+     * Allocating 0 pages is not allowed.
+     */
+    assert(n >= 1);
+
+    while (1) {
+        /*
+         * Look for the first free page starting at (b), initially 0.
+         */
+        a = ffs_at(alloc->bmap, alloc->bmap_size, b);
+        if (a < 0)
+            return NULL;
+        /*
+         * Look for the first used page after the found free page.
+         */
+        b = ffc_at(alloc->bmap, alloc->bmap_size, a);
+        if (b < 0)
+            /*
+             * Nothing found; all remaining pages from a..bmap_bits are free.
+             */
+            b = bmap_bits;
+        /*
+         * Is the block big enough? If yes, mark as used (0) and return it.
+         */
+        if (b - a >= n) {
+            clearn_at(alloc->bmap, alloc->bmap_size, a, n);
+            return (void *)(alloc->start_addr + (a * PAGE_SIZE));
+        }
+        /*
+         * Stop the search if we hit the end of bmap[] and did not find a large
+         * enough block.
+         */
+        if (b == bmap_bits)
+            return NULL;
+        /*
+         * If we got here, loop with (b) set to the last seen used page.
+         */
+    }
+}
+
+/*
+ * Free (n) pages at (addr) from (alloc).
+ */
+void bmap_free(bmap_allocator_t *alloc, void *addr, size_t n)
+{
+    /*
+     * Verify that:
+     *    addr is page-aligned
+     *    addr is within the range given to alloc
+     *    n is at least 1; the maxiumum size of n is checked in clearn_at().
+     */
+    assert(((uintptr_t)addr & (PAGE_SIZE - 1)) == 0);
+    assert((uintptr_t)addr >= alloc->start_addr);
+    assert(n >= 1);
+
+    int a = ((uintptr_t)addr - alloc->start_addr) / PAGE_SIZE;
+    setn_at(alloc->bmap, alloc->bmap_size, a, n);
```

Comment:
this sets n bits to 1 starting at a, is this intentional? my expectation was that `clearn_at` would be called here.

Reply:
ah no, sorry - free is 1, used is 0. thus setting the bits to 1 is marking them free.